feat(sync/koreader): Add KOReader reading progress synchronization

zeedif opened the pull request:
This pull request introduces a comprehensive integration with KOReader Sync Server, enabling two-way synchronization of reading progress between Suwayomi-Server and KOReader devices. This allows users to maintain a consistent reading experience across platforms.

#### Key Features

*   **Two-Way Synchronization:** Reading progress (last page read and read status) is pushed to the KOReader Sync Server when updated in Suwayomi and pulled when a chapter is accessed, ensuring the latest progress is always available.
*   **Configurable Sync Strategy:** Users can define how conflicts are handled via the `koreaderSyncStrategy` setting, with options to prioritize the latest timestamp (`LATEST`), always trust the KOReader server (`KOSYNC`), or always trust Suwayomi-Server (`SUWAYOMI`).
*   **GraphQL API for Sync Management:** A new set of GraphQL queries and mutations has been added to manage the KOReader Sync connection seamlessly:
    *   `koSyncStatus`: Checks the current login status.
    *   `connectKoSyncAccount`: A simplified flow that attempts to log in and, if the user doesn't exist, automatically registers them on the sync server.
    *   `logoutKoSyncAccount`: Clears local credentials to disconnect from the service.
*   **OPDS Integration:** Chapter metadata feeds now automatically pull the latest progress from the KOReader Sync Server, providing up-to-date reading status to OPDS clients.
*   **Settings and Backup:** All new KOReader Sync settings are fully integrated into the server configuration, GraphQL settings API, and the backup/restore system.

#### Technical Implementation

*   A new `KoreaderSyncService` has been implemented to encapsulate all logic related to authentication, hashing, and communication with the sync server.
*   The `Chapter` table has been extended with a `koreader_hash` column to uniquely identify chapter files (CBZs) on the sync server. A database migration (`M0051`) is included.
*   CBZ files are now hashed on download completion using KOReader's specific algorithm to ensure document compatibility.
*   Progress updates via the API (`ChapterMutation`) and OPDS (`OpdsEntryBuilder`) now trigger sync operations.
*   Configuration options for the feature have been refactored to use enums (`KoreaderSyncStrategy`, `KoreaderSyncChecksumMethod`) for improved type safety and robustness.

#### For Frontend (WebUI) Developers

To ensure the reader opens on the correct, synchronized page, a small adjustment is required in the WebUI's GraphQL query for fetching chapter pages. The `fetchChapterPages` mutation now returns the updated progress fields on the `chapter` object after performing a sync check.

**Before:**

```graphql
mutation GET_CHAPTER_PAGES_FETCH($input: FetchChapterPagesInput!) {
  fetchChapterPages(input: $input) {
    chapter {
      id
      pageCount
      __typename
    }
    pages
    __typename
  }
}
```

**After:**

The query should be updated to request `lastPageRead` and other relevant progress fields.

```graphql
mutation GET_CHAPTER_PAGES_FETCH($input: FetchChapterPagesInput!) {
  fetchChapterPages(input: $input) {
    chapter {
      id
      pageCount
      lastPageRead  # <-- Added: The synced last page read
      lastReadAt    # <-- Recommended: The timestamp of the last read
      isRead        # <-- Recommended: The updated read status
      __typename
    }
    pages
    __typename
  }
}
```

By making this change, the WebUI will automatically receive the most up-to-date reading progress when a user opens a chapter, allowing the reader to jump to the correct page seamlessly.

zeedif committed:
feat(sync/koreader): implement reading progress synchronization

This commit introduces a comprehensive integration with KOReader Sync Server to enable two-way synchronization of reading progress.

The core logic is encapsulated in a new `KoreaderSyncService`, which handles authentication, registration, and progress pushing/pulling based on user-defined strategies (LATEST, KOSYNC, SUWAYOMI).

Key changes include:

- A new GraphQL API is added to manage the KOReader Sync connection:
  - `connectKoSyncAccount` mutation provides a simplified flow that attempts to log in and, if the user doesn't exist, automatically registers them.
  - `logoutKoSyncAccount` mutation to clear credentials.
  - `koSyncStatus` query to check the current connection status.

- Reading progress is now synchronized at key points:
  - The `fetchChapterPages` mutation pulls the latest progress from the sync server before loading the reader. It respects the configured sync strategy and updates the local database if necessary.
  - The `updateChapters` and other progress-updating methods now push changes to the sync server automatically.
- OPDS chapter entries also pull the latest progress, ensuring clients receive up-to-date reading status.

- Supporting backend changes have been made:
  - The `Chapter` table is extended with a `koreader_hash` column to uniquely identify documents. A database migration is included.
  - New configuration options are added to `server.conf` to manage the feature (enable, server URL, credentials, strategy, etc.).

zeedif committed:
perf(opds): defer KOReader sync to improve chapter feed performance

Removes the KOReader Sync progress-pulling logic from the `createChapterListEntry` function.

The previous implementation triggered a network request to the sync server for every single chapter when generating a list, leading to severe performance issues and slow load times on feeds with many entries.

This change reverts to the more performant approach of always linking to the chapter's metadata feed. The expensive sync operation will be handled within the metadata entry generation instead, ensuring it's only triggered on-demand for a single chapter. This restores the responsiveness of browsing chapter feeds.

zeedif committed:
refactor(koreader): Use enums for sync settings and correct OPDS logic

Refactor Koreader Sync settings to use enums instead of raw strings for `checksumMethod` and `strategy`. This improves type safety, prevents typos, and makes the configuration handling more robust.

The changes include:
- Introducing `KoreaderSyncChecksumMethod` and `KoreaderSyncStrategy` enums.
- Updating `ServerConfig`, GraphQL types, and backup models to use these new enums.
- Refactoring `KoreaderSyncService` to work with the enum types.

Additionally, this commit fixes an issue in `OpdsEntryBuilder` where the logic for determining which sync progress to use (local vs. remote) was duplicated. The builder now correctly delegates this decision to `KoreaderSyncService.pullProgress`, which already contains the necessary strategy logic. This centralizes the logic and ensures consistent behavior.

Syer10 commented on server/src/main/kotlin/suwayomi/tachidesk/manga/impl/sync/KoreaderSyncService.kt:
```diff
@@ -0,0 +1,359 @@
+            }
+
+        return try {
+            network.client.newCall(request).execute().use { response ->
```
This needs to use .await() instead of .execute(), make sure to change each instance

Syer10 commented on server/src/main/kotlin/suwayomi/tachidesk/manga/impl/sync/KoreaderSyncService.kt:
```diff
@@ -0,0 +1,359 @@
+
+    suspend fun logout() {
+        GlobalConfigManager.updateValue("server.koreaderSyncUsername", "")
+        GlobalConfigManager.updateValue("server.koreaderSyncUserkey", "")
```
This needs to set the values in serverConfig, not GlobalConfigManager
```suggestion
        serverConfig.koreaderSyncUsername.value = ""
        serverConfig.koreaderSyncUserkey.value = ""
```

Syer10 commented on server/src/main/kotlin/suwayomi/tachidesk/manga/impl/sync/KoreaderSyncService.kt:
```diff
@@ -0,0 +1,359 @@
+            val registerResult = register(username, userkey)
+            return if (registerResult.success) {
+                GlobalConfigManager.updateValue("server.koreaderSyncUsername", username)
+                GlobalConfigManager.updateValue("server.koreaderSyncUserkey", userkey)
```
```suggestion
                serverConfig.koreaderSyncUsername.value = username
                serverConfig.koreaderSyncUserkey.value = userkey
```

Syer10 commented on server/src/main/kotlin/suwayomi/tachidesk/manga/impl/sync/KoreaderSyncService.kt:
```diff
@@ -0,0 +1,359 @@
+
+        if (authResult.success) {
+            GlobalConfigManager.updateValue("server.koreaderSyncUsername", username)
+            GlobalConfigManager.updateValue("server.koreaderSyncUserkey", userkey)
```
```suggestion
           serverConfig.koreaderSyncUsername.value = username
           serverConfig.koreaderSyncUserkey.value = userkey
```

Syer10 reviewed (changes requested):
A quick review, will do more later.

This will likely have to be merged in a week or 2, as I will be unavailable for the next week.

zeedif committed:
refactor(koreader): Improve config handling and remove redundant update

This commit combines several refactoring and cleanup tasks:

- **Koreader Sync:** The sync service is updated to use the modern `serverConfig` provider instead of the legacy `GlobalConfigManager`. This aligns it with the current configuration management approach in the project.

- **Download Provider:** A redundant `pageCount` database update is removed from `ChaptersFilesProvider`. This operation was unnecessary because the `getChapterDownloadReady` function, which is called earlier in the download process, already verifies and corrects the page count. This change eliminates a superfluous database write and fixes a related import issue.

Syer10 commented on server/src/main/kotlin/suwayomi/tachidesk/manga/impl/download/fileProvider/ChaptersFilesProvider.kt:
```diff
@@ -105,27 +107,6 @@ abstract class ChaptersFilesProvider<Type : FileType>(
         scope: CoroutineScope,
         step: suspend (DownloadChapter?, Boolean) -> Unit,
     ): Boolean {
-        val existingDownloadPageCount =
```
Looks like you accidentally reverted one of the recent commits 😄

Syer10 commented on server/src/main/kotlin/suwayomi/tachidesk/manga/impl/download/fileProvider/ChaptersFilesProvider.kt:
```diff
@@ -189,6 +191,25 @@ abstract class ChaptersFilesProvider<Type : FileType>(
+            ChapterTable.update({ ChapterTable.id eq chapterId }) {
+                it[ChapterTable.pageCount] = getImageCount()
+            }
+        }
```
This can be removed, as we already do this in `ChapterForDownload.getChapterDownloadReady`

zeedif committed:
feat(sync/koreader)!: enhance sync strategy and add progress tolerance

This commit overhauls the KOReader synchronization feature to provide more granular control and robustness. The simple on/off toggle has been replaced with a more flexible strategy-based system.

Key changes include:
- Replaced `koreaderSyncEnabled` with a more powerful `koreaderSyncStrategy` enum.
- Introduced new sync strategies: `PROMPT`, `SILENT`, `SEND`, `RECEIVE`, and `DISABLE`, allowing for fine-grained control over the sync direction and conflict resolution.
- Added a `koreaderSyncPercentageTolerance` setting. This prevents unnecessary sync updates for minor progress differences between Suwayomi and KOReader.
- Refactored the `KoreaderSyncService` to implement the new strategies and use the configurable tolerance.
- Updated GraphQL schemas, mutations, and server configuration to remove the old setting and incorporate the new ones.
- Adjusted the backup and restore process to correctly handle the new configuration parameters.
- Modified API endpoints and internal logic to check and apply remote progress based on the selected strategy.

BREAKING CHANGE: The `koreaderSyncEnabled` setting is removed and replaced by a more granular `koreaderSyncStrategy`. The enum values for the strategy have been completely changed, making previous configurations incompatible.

zeedif committed:
fix: remove unused imports

Syer10 reviewed (approved):
Seems good to me. I'l look at the launcher and docker container soon

schroda commented on server/src/main/kotlin/suwayomi/tachidesk/graphql/mutations/ChapterMutation.kt:
```diff
@@ -248,9 +263,28 @@ class ChapterMutation {
         return future {
-            asDataFetcherResult {
-                val chapter = getChapterDownloadReadyById(chapterId)
+            var chapter = getChapterDownloadReadyById(chapterId)
```
Mutations need to be wrapped in `asDataFetcherResult` to handle partial successful mutation responses.
Otherwise, in case a request includes multiple mutations, one failing mutation would lead to all mutations to not return any data in the request

schroda commented on server/src/main/resources/server-reference.conf:
```diff
@@ -85,3 +85,12 @@ server.opdsMarkAsReadOnDownload = false
+server.koreaderSyncUserkey = ""
+server.koreaderSyncDeviceId = ""
+server.koreaderSyncChecksumMethod = "binary" # "binary" or "filename"
+server.koreaderSyncStrategy = "latest" # "latest", "suwayomi", "kosync"
```
This looks to be outdated, `latest` does not exist in the enum.
It probably also should be `disable` by default, otherwise, the koreader sync will kick in and fail for every user, regardless of them using koreader or not?

zeedif replied:
The sync logic is gated by the koreaderSyncUsername and koreaderSyncUserkey fields not being empty, so the strategy's default value doesn't affect users who haven't configured them. Plus, the connect mutation already sets the strategy to prompt on a successful login. So, the default value mainly serves as a safeguard for users who might set their username and userkey directly via environment variables, bypassing the login flow.

schroda replied:
> The sync logic is gated by the koreaderSyncUsername and koreaderSyncUserkey fields not being empty, so the strategy's default value doesn't affect users who haven't configured them.

I see. Looks like I missed that.

>  Plus, the connect mutation already sets the strategy to prompt on a successful login

What's the reasoning behind that? I would assume that this will get reported as a bug in the future.

If a user changed the sync strategy and at some point logged out and back in again, this would just randomly change the previously set sync strategy without the user realizing that it happened.

Additionally, it looks like the connect mutation doesn't inform the client about it mutating this setting. Which would lead to an out-of-date cache on the client.

schroda replied:
> Additionally, it looks like the connect mutation doesn't inform the client about it mutating this setting. Which would lead to an out-of-date cache on the client.

I just realized that the connect mutation mutates the server settings but does not return the SettingType.
Which means that the client has to manually update its cache (username and password) instead of it happening automatically through the returned type

schroda commented on server/src/main/resources/server-reference.conf:
```diff
@@ -85,3 +85,12 @@ server.opdsMarkAsReadOnDownload = false
 server.opdsChapterSortOrder = "DESC" # "ASC", "DESC"
+
+# Koreader Sync
+server.koreaderSyncEnabled = false
```
This does not get used anywhere

zeedif committed:
feat(opds, sync): enhance Koreader sync and OPDS conflict handling

This commit introduces significant improvements to the Koreader synchronization feature, focusing on providing a better user experience for handling progress conflicts in both OPDS and GraphQL clients.

Key changes include:

- **OPDS Conflict Resolution:** When a reading progress conflict is detected, the OPDS feed for a chapter now provides two distinct "Read Online" links: one to continue from the local progress and another to continue from the synced progress from the remote device (e.g., "Continue Reading Online (Synced from KOReader)"). This empowers users to choose which progress to follow.

- **GraphQL Sync Conflict Information:** The `fetchChapterPages` GraphQL mutation now includes a `syncConflict` field in its payload. This field provides the remote device name and page number, allowing GraphQL clients to implement a user-facing prompt to resolve sync conflicts.

- **Improved Sync Strategy Handling:**
  - The `connectKoSyncAccount` mutation no longer unconditionally sets the sync strategy to `PROMPT`. It now respects the user's existing setting, preventing accidental configuration changes upon re-login.
  - The default `koreaderSyncStrategy` in the configuration is changed to `DISABLED`, providing a safer and more intuitive default for new users.

- **Refinements & Fixes:**
  - The fallback for the remote device name is now centralized within the KoreaderSyncService, defaulting to "KOReader" if not provided.
  - Renamed `KoreaderSyncStrategy.DISABLE` to `DISABLED` for consistency.
  - Updated i18n strings for OPDS links to be more descriptive and user-friendly.

zeedif committed:
refactor(kosync): rename stream page link titles for consistency

zeedif committed:
refactor(kosync): return SettingsType in auth mutation payloads

The `connectKoSyncAccount` and `logoutKoSyncAccount` mutations modify server settings (username and userkey) but did not previously return the updated configuration. This forced client applications to manually refetch settings to avoid a stale cache.

This change modifies the payloads for both mutations to include the full `SettingsType`.

By returning the updated settings directly, GraphQL clients like Apollo Client can automatically update their cache, simplifying client-side state management and ensuring the UI always reflects the current server configuration.

Additionally, `clientMutationId` has been added to `KoSyncConnectPayload` for consistency with GraphQL practices, aligning it with the logout mutation.

Refs: #1560

zeedif committed:
refactor(kosync): replace KoSyncConnectPayload with ConnectResult in connect method

Syer10 commented on server/src/main/resources/server-reference.conf:
```diff
@@ -85,3 +85,11 @@ server.opdsMarkAsReadOnDownload = false
+server.koreaderSyncUserkey = ""
+server.koreaderSyncDeviceId = ""
+server.koreaderSyncChecksumMethod = "binary" # "binary" or "filename"
+server.koreaderSyncStrategy = "disabled" # "prompt", "silent", "send", "receive", "disabled"
```
This is missing the server.koreaderSyncPercentageTolerance configuration

zeedif committed:
fix(kosync): add koreaderSyncPercentageTolerance default setting
